Programming Fundamentals Assignment 1

meta-akshita-dixit opened the pull request:
Sir,this is the assignment number 1 having 6 questions. Kindly review it and let me know if any changes are to me made.
Thank you.

meta-akshita-dixit committed:
Programming Fundamentals Assignment 1

metacube-shivang commented on PF-session1/assignment1/BinToOctal.java:
```diff
@@ -0,0 +1,46 @@
+  int convertBinaryToOctal(int n)               
+	{
+		int octal=0,i=0;
+		while(n!=0)
```
What if n is negative?

metacube-shivang commented on PF-session1/assignment1/TestBinToOct.java:
```diff
@@ -0,0 +1,23 @@
+
+
+public class TestBinToOct {
+	BinToOctal test =new BinToOctal();
```
Different methods in a unit test should run independently. For this, normally, we do not instantiate global variables in a unit test class.

metacube-shivang commented on PF-session1/assignment1/BinToOctal.java:
```diff
@@ -0,0 +1,46 @@
+	 * @param binary
+	 * @return
+	 */
+   int toDecimal(int binary)                      
```
The methods do not rely on any class variable. So, why not make them static?

metacube-shivang commented on PF-session1/assignment1/BinToOctal.java:
```diff
@@ -0,0 +1,46 @@
+	{
+		int decimal=0,i=0;
+		while(binary!=0)
+		{
```
Use recommended syntax for where to open and close braces.

metacube-shivang commented on PF-session1/assignment1/TestBinToOct.java:
```diff
@@ -0,0 +1,23 @@
+public class TestBinToOct {
+	BinToOctal test =new BinToOctal();
+	@Test
+	public void test() {
```
Test methods should be named such that it gives you an idea of what it is testing.  For example: this method could be "testConversionSuccess()" to indicate that it is testing for a positive case.

metacube-shivang commented on PF-session1/assignment1/TestBinToOct.java:
```diff
@@ -0,0 +1,23 @@
+	}
+	@Test
+	public void test2() {                                            //false test case
+		assertEquals(1,test.convertBinaryToOctal(100001));
```
This test will fail. The test cases always need to pass, even when they are testing negative test cases. In this case, you should use "assertNotEquals". That way it tests that the returned value is not 1 and also the unit test passes.

metacube-shivang commented on PF-session1/assignment2/RemovingDuplicates.java:
```diff
@@ -0,0 +1,48 @@
+		{
+			if(input[i] == input[loopCounter])           //compares with each element of the array.
+			{
+				flag = false;
```
If a duplicate is detected, no need to check the remainder of the array; simply return false the first time the duplicate is detected.

metacube-shivang commented on PF-session1/assignment2/RemovingDuplicates.java:
```diff
@@ -0,0 +1,48 @@
+			else
+			{
+				for(int loopCounter = i ; loopCounter < size-1; loopCounter++)
+					input[loopCounter] = input[loopCounter + 1];
```
We are shifting all of the elements to the left, one element at a time. Consider a case where the input array is a large array consisting of only one unique element. Say, an array with a thousand 1s.

In this case, you'll be unnecessarily shifting elements repeatedly, even though ultimately the resulting array only needs to contain a single 1.

Think about an alternate way to handle this, without copying over duplicate elements.

metacube-shivang commented on PF-session1/assignment2/TestRemovingDuplicate.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.junit.Test;
+
+public class TestRemovingDuplicate {
```
Follow other suggestions from assignment 1 regarding naming, test coverage, syntax, etc.

metacube-shivang commented on PF-session1/assignment3/LongestSequence.java:
```diff
@@ -0,0 +1,40 @@
+			}
+			counter=1;
+		}
+		return Arrays.copyOfRange(input, position, position + count);
```
What if array is empty?

metacube-shivang commented on PF-session1/assignment3/TestLongestSequence.java:
```diff
@@ -0,0 +1,25 @@
+{
+	LongestSequence test;
+	int arr[],expectedArr[];
+	@Before
```
@Before is useful when setting up common data between multiple unit tests. Here, there's only one unit test. So why do we need @Before?

Also, note that even if there are multiple unit tests, use of @Before will fail in this case because unit tests could potentially be run in multi-threaded environment, which will cause problems in our test because we are instantiating common data structures (arr[] and expectedArr[]) in the "before" method.

metacube-shivang commented on PF-session1/assignment3/TestLongestSequence.java:
```diff
@@ -0,0 +1,25 @@
+		assertArrayEquals(expectedArr,test.longestSequence(arr));
+	}
+
+}
```
Add more test cases (empty array, array with single element, array with negative values, with all elements equal, etc.)

metacube-shivang commented on PF-session1/assignment4/FirstComeFirstServe.java:
```diff
@@ -0,0 +1,42 @@
+			}
+			else
+			{
+				outputArr[loopCounter][2] = 0;                                                           //waiting time
```
Setting of outputArr[][2] outputArr[][3] and outputArr[][4] is being done repeatedly at three places. This code should be extracted to a common method, with just required values being passed to it.

metacube-shivang commented on PF-session1/assignment5/AscOrDesc.java:
```diff
@@ -0,0 +1,38 @@
+				for(position=1; position < array.length - 1; position++)
+				{
+					if(array[position] > array[position + 1])
+						type=0;
```
if type = 0 is detected, we can return immediately; no need to check for remaining elements.

metacube-shivang commented on PF-session1/assignment5/AscOrDesc.java:
```diff
@@ -0,0 +1,38 @@
+				type=2;
+				for(position = 1; position < array.length - 1; position++)
+				{
+					if(array[position] <= array[position +1 ])
```
why are we comparing with "<=" instead of just "<"?

metacube-shivang commented on PF-session1/assignment5/TestAscOrDesc.java:
```diff
@@ -0,0 +1,57 @@
+		assertEquals(2,checkSort.ascOrDesc(arr1));
+	}
+	
+	@Before
```
Each @Before runs before each @Test, and there's no guarantee which will run in which order. What do you think the behavior will be when there are two @Before in the same test class? :)

metacube-shivang reviewed (changes requested):
Hi @meta-akshita-dixit,

I've added some review comments. There are some that are problem specific, and some that are generic.

The generic ones have only been added to problem 1, but should be implemented in all the problems.

If there are any questions, we can discuss them tomorrow.

meta-akshita-dixit committed:
Updated

Removed the global variables
Removed the use of annotation before
added more test cases

meta-akshita-dixit committed:
Updates

Added the condition if array is empty

meta-akshita-dixit committed:
Updates

generic changes made

meta-akshita-dixit committed:
updates

value returned as 0 encountered

meta-akshita-dixit committed:
Updates

-generic changes made

meta-akshita-dixit committed:
Update TestFCFS.java

meta-akshita-dixit committed:
updates

-generic changes
-negative argument test case

meta-akshita-dixit committed:
update

-negative value of n
-static method

meta-akshita-dixit committed:
Updates

-returned false as duplicate encountered for first time
-shifting loop removed

metacube-shivang commented on PF-session1/assignment3/TestLongestSequence.java:
```diff
@@ -7,19 +7,54 @@
+		expectedArr=new int[]{-3,-2,0,1};
+		assertArrayEquals(expectedArr,test.longestSequence(arr));
+	}
 }
```
Much better test cases! 👍
